content/blog/2020-12-16-compcert/index.md

```diff
@@ -0,0 +1,249 @@
+
+The __third pass__ translates Clight into the [C#minor][csminor] intermediate representation, simplifying loop and `switch` statements in addition to resolving address computation and overloaded function applications. 
+
+The __fourth pass__ translates C#minor into Cminor, where the only addressable variable a function takes is the stack. Cminor is the last intermediate language in the compiler that is independent of computer architecture. Work on [separation logic for Cminor][cmsl] to support reasoning about concurrency as well as the development of a [verified static analyzer for C#minor][csmsa] make Cminor an attractive target for other C compilers. 
+
+The __fifth pass__ translates Cminor into CminorSel, which performs instruction selection on sequences of operations that can be done in one instruction (for instance, an add instruction with a constant becomes an immediate-add instruction for 32-bit Intel architecture). Loads and stores are also translated into the target architecture's addressing conventions. The correctness theorem of this pass does not hold if the target language doesn't support required operations.
+
+The __sixth pass__ generates a CFG from CminorSel with nodes in [RTL][rtl]; it looks like a [graph of instructions][rtlv]. Included is a size-based heuristic to select which branch of an `if` statement is the fallthrough one. 
+
+Now, we can start optimizing the RTL code. The following passes transform the RTL CFG.
+
+The __seventh pass__ transforms ignored or immediately used function calls into tail calls, which do not need a new stack frame to be allocated. This transformation can remove transitions in the CFG, such as returning a function call result directly instead of a variable assigned to it and used only to return the result. That'll throw a wrench in our proof of correctness, but more on that later. 
+
+The __eighth pass__ performs function inlining based on the presence of the `inline` keyword at the call site. Recursive functions are inlined once, and recursive calls are not inlined (especially since inlining a function that increases the caller's stack block beyond the largest integer causes compilation to fail). Since this pass can eliminate CFG nodes, proving correctness needs to be adjusted.  
+
+The __ninth pass__ renumbers the CFG nodes in post-order with respect to their edges, in order to make the following static analyses more efficient. The algorithm used is Kildall's, outlined in pseudocode [here][kildall]. 
+
+The __tenth pass__ performs constant propagation. This pass may also replace instruction operators in the case that some arguments are statically known. Since the number of transitions may change, correctness needs to take this into account. 
+
+The __eleventh pass__ executes common subexpression elimination, although it's not performed on function calls and it doesn't consider `x + y` and `y + x` equivalent. 
```

Comment:
Not important at all, but: any particular reason for not doing the commutativity thing?

Reply:
Looks like this is actually not true in the current implementation. I will update the document to reflect that this is a thing of the past!

Reply:
It's important! I think this was the case in the original paper's implementation due to the way they were solving the dataflow equations, but has been amended by strengthening their LVN (!) in later versions. I will update this section to reflect that.